### PR TITLE
User management

### PR DESCRIPTION
**Pull request by Bronson-Brown-deVost**
The user management endpoints are written and tested.

Part of the user management uses an emailer.  I use Gmail for this right now until we know what will be used long term.  For that functionality to work, you will need to provide the email credentials as environment variables.  Set your gmail username with SQE_GMAIL_USERNAME (i.e., the name part of name@gmail.com), and your password with SQE_GMAIL_PASSWORD (probably you can set these environment variables in visual studio).  You must also make sure your gmail account allows less secure apps (https://support.google.com/accounts/answer/6010255?hl=en).

### CODE REVIEW

File: sqe-http-api/Controllers/ArtefactController.cs [hunk 1]
```diff
@@ -99,10 +99,10 @@ public async Task<ActionResult<ArtefactDTO>> GetArtefact([FromRoute] uint editio
         }
         
         /// <summary>
-        /// Provides a listing of all artefacts that are part of the specified edition, including data about
-        /// the related images
+        /// Creates a new artefact with the provided data.
```
Comment (Bronson-Brown-deVost) [inline sqe-http-api/Controllers/ArtefactController.cs:111]: I guess I fixed some documentation here.

File: sqe-http-api/Controllers/ArtefactController.cs [hunk 2]
```diff
@@ -124,19 +124,5 @@ public async Task<ActionResult<ArtefactDTO>> GetArtefact([FromRoute] uint editio
                 return NotFound();
             }
         }
-
```
Comment (Bronson-Brown-deVost) [inline sqe-http-api/Controllers/ArtefactController.cs:127]: I removed these, which are no longer used.

File: sqe-http-api/DTOs/Artefact.cs [hunk 3]
```diff
@@ -38,4 +38,19 @@ public class ArtefactDesignationListDTO
     {
         public List<ArtefactDesignationDTO> artefactDesignations { get; set; }
     }
+
```
Comment (Bronson-Brown-deVost) [inline sqe-http-api/DTOs/Artefact.cs:41]: I put these DTOs in the proper place.

File: data-access/data-access.csproj [hunk 4]
```diff
@@ -14,4 +14,10 @@
     <PackageReference Include="Microsoft.Extensions.Configuration.Abstractions" Version="2.2.0" />
     <PackageReference Include="Wkx" Version="0.5.0" /> 
   </ItemGroup>
+
```
Comment (Bronson-Brown-deVost) [inline data-access/data-access.csproj]: I am not really sure why this is here.

File: data-access/data-access.csproj [hunk 5]
```diff
@@ -14,4 +14,10 @@
...
+  <ItemGroup>
+    <Reference Include="Microsoft.Extensions.Options, Version=2.2.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60">
+      <HintPath>..\..\.nuget\packages\microsoft.extensions.options\2.2.0\lib\netstandard2.0\Microsoft.Extensions.Options.dll</HintPath>
+    </Reference>
+  </ItemGroup>
```
Comment (Bronson-Brown-deVost) [inline data-access/data-access.csproj]: I am not really sure why this is here.

Comment (Bronson-Brown-deVost) [reply data-access/data-access.csproj]: It is unnecessary, I have removed it.

File: data-access/Queries/User.cs [hunk 6]
```diff
@@ -37,4 +52,139 @@ public class Return
...
+    /// Checks whether an account with @Username or @Email exists and whether it is authenticated yet.
+    /// </summary>
+    internal static class CheckUserAuthentication
+    {
+        public const string GetQuery = @"
+SELECT user_id, authenticated, user_name, email
+FROM user
+WHERE user_name = @Username OR email = @Email";
+
+        public class Return
```
Comment (zmbq) [inline data-access/Queries/User.cs]: Don't we usually call this "Result"?

File: data-access/Queries/User.cs [hunk 7]
```diff
@@ -37,4 +52,139 @@ public class Return
...
+        {
+            public int user_id { get; set; } 
+            public bool authenticated { get; set; }
```
Comment (zmbq) [inline data-access/Queries/User.cs]: "authenticated" is the wrong term. An Authenticated user means a user that has logged in. This has confused me throughout the code. We should "activated" - users activate their accounts.

File: data-access/Queries/User.cs [hunk 8]
```diff
@@ -37,4 +52,139 @@ public class Return
...
+    /// <summary>
+    /// Creates an entry in the user_email_token table for @UserId with the token @Token for the request type @Type
+    /// (CreateUserEmailTokenQuery.Activate or CreateUserEmailTokenQuery.ResetPassword).
+    /// </summary>
+    internal static class CreateUserEmailTokenQuery
+    {
+        public const string GetQuery = @"
+INSERT INTO user_email_token (user_id, token, type)
+VALUES(@UserId, @Token, @Type)";
+
```
Comment (zmbq) [inline data-access/Queries/User.cs:119]: We should add an expiration date for tokens. A token can't be used if it has expired.

File: data-access/UserRepository.cs [hunk 9]
```diff
@@ -11,14 +15,33 @@ public interface IUserRepository
...
+        /// <summary>
+        /// Returns user information based on the submitted credentials.
+        /// </summary>
+        /// <param name="userName">The user's username</param>
+        /// <param name="password">The user's password</param>
+        /// <returns></returns>
         public async Task<User> GetUserByPasswordAsync(string userName, string password)
         {
-            var sql = @"SELECT user_name, user_id FROM user WHERE user_name =@UserName AND pw = SHA2(@Password, 224)";
+            var sql = @"SELECT user_name, user_id FROM user WHERE user_name = @UserName AND pw = SHA2(@Password, 224) AND authenticated = 1";
```
Comment (zmbq) [inline data-access/UserRepository.cs]: This probably means users who haven't activated their accounts cannot log in. This is not ideal. Users with inactive accounts *should* be able to log in with their password, and then see a message asking them to activate their account. They will also have a link for resending the activation email.

File: data-access/UserRepository.cs [hunk 10]
```diff
@@ -44,5 +73,262 @@ public async Task<UserEditionPermissions> GetUserEditionPermissionsAsync(uint us
...
+        }
+        
+        /// <summary>
+        /// Creates a request in the database for a reset password token.
+        /// </summary>
+        /// <param name="email">Email address of the user who has forgotten the password</param>
+        /// <returns>Returns user information and secret token that are used to format the reset password email</returns>
+        public async Task<User> RequestResetForgottenPasswordAsync(string email)
+        {
+            await CleanOldTokens(); // Always flush the expired tokens from the database first
```
Comment (zmbq) [inline data-access/UserRepository.cs]: This function should be called `CleanOldTokensAsync`.

File: data-access/UserRepository.cs [hunk 11]
```diff
@@ -44,5 +73,262 @@ public async Task<UserEditionPermissions> GetUserEditionPermissionsAsync(uint us
...
+
+        /// <summary>
+        /// Clears out old tokens, but see the comments to DeleteOldTokens.
+        /// </summary>
+        /// <returns></returns>
+        private async Task CleanOldTokens()
+        {
+            using (var connection = OpenConnection())
+            {
+                await connection.ExecuteAsync(DeleteOldTokens.GetQuery, new { Days = _tokenValidDays });
```
Comment (zmbq) [inline data-access/UserRepository.cs]: I see now that tokens do have expiry dates, but I couldn't see them in the queries.

File: sqe-http-api/Controllers/UserController.cs [hunk 12]
```diff
@@ -20,34 +23,126 @@ public UserController(IUserService userServiceAuthenticate)
...
+            {
+                return Conflict(new {message = err.Message}); // I would rather not tell which element is wrong (again someone might phish for valid emails), but maybe it is not a problem.
+            }
+        }
+        
+        /// <summary>
+        /// Confirms creation of new user account.
+        /// </summary>
+        /// <param name="payload">JSON object with token from user registration email.</param>
+        /// <returns>Returns 200 for success and 404 for failure.</returns>
```
Comment (zmbq) [inline sqe-http-api/Controllers/UserController.cs]: If the result is empty, we should return a 204.

File: sqe-http-api/Controllers/UserController.cs [hunk 13]
```diff
@@ -20,34 +23,126 @@ public UserController(IUserService userServiceAuthenticate)
...
         /// <param Name="userParam">JSON object with a username and password parameter</param>
         [AllowAnonymous]
-        [HttpPost("login")] // api/v1/user/login
+        [HttpPost("login")]
         public async Task<ActionResult<LoginResponseDTO>> AuthenticateAsync([FromBody]LoginRequestDTO userParam)
         {
             var user = await _userService.AuthenticateAsync(userParam.userName, userParam.password);
             if (user == null)
-                return Unauthorized(new { message = "Username or password is incorrect", code = 600}); // TODO: Add Error Code
+                return Unauthorized(new { message = "Username or password is incorrect ", code = 600}); // TODO: Add Error Code
```
Comment (zmbq) [inline sqe-http-api/Controllers/UserController.cs]: You added an unnecessary space. Also, the error code has already been added, so you can delete the TODO

This serves as the petty review comment of the day

File: sqe-http-api/Controllers/UserController.cs [hunk 14]
```diff
@@ -20,34 +23,126 @@ public UserController(IUserService userServiceAuthenticate)
...
-            return user;
+            await _userService.RequestResetLostPasswordAsync(payload.email);
+            return NoContent();
+        }
+        
+        /// <summary>
+        /// Change the password for the currently logged in user.
+        /// </summary>
+        /// <param name="payload">A JSON object with the old password and the new password.</param>
+        /// <returns>Status 200 with a successful request or status 409 for an unsuccessful request</returns>
```
Comment (zmbq) [inline sqe-http-api/Controllers/UserController.cs]: Again, return 204 for empty successful queries

File: sqe-http-api/DTOs/User.cs [hunk 15]
```diff
@@ -22,4 +22,54 @@ public class LoginResponseDTO : UserDTO
     {
         public string token { get; set; }
     }
+    
+    public class NewUserDTO
```
Comment (zmbq) [inline sqe-http-api/DTOs/User.cs]: Why is this different from the regular UserDTO?

File: sqe-http-api/DTOs/User.cs [hunk 16]
```diff
@@ -22,4 +22,54 @@ public class LoginResponseDTO : UserDTO
...
+        /// <param name="email">Email address for the new user account, must be unique</param>
+        /// <param name="password">Password for the new user account</param>
+        /// <param name="organization">Name of affiliated organization (if any)</param>
+        /// <param name="forename">The user's given name (may be empty)</param>
+        /// <param name="surname">The user's family name (may be empty)</param>
+        public NewUserDTO(string userName, string email, string password, string organization, string forename, string surname)
+        {
+            this.userName = userName;
+            this.email = email;
+            this.password = password;
```
Comment (zmbq) [inline sqe-http-api/DTOs/User.cs]: Don't ever ever return passwords from the backend back to the frontend.

File: sqe-http-api/Helpers/SystemEmailer.cs [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+    public class DevEmailSender : IEmailSender
+    {
+        /// <summary>
+        /// Sends an email
+        /// </summary>
+        /// <param name="email">Email address to send the message to.</param>
+        /// <param name="subject"></param>
+        /// <param name="htmlMessage"></param>
+        /// <returns></returns>
+        public async Task SendEmailAsync(string email, string subject, string htmlMessage)
```
Comment (zmbq) [inline sqe-http-api/Helpers/SystemEmailer.cs:22]: Do we want some code that takes Razor templates and turns them into emails? Or is this overkill?

File: sqe-http-api/Services/UserService.cs [hunk 18]
```diff
@@ -19,28 +19,40 @@ public interface IUserService
         UserDTO GetCurrentUser();
         uint? GetCurrentUserId();
         UserInfo GetCurrentUserObject(uint? editionId = null);
+        Task<UserDTO> CreateNewUserAsync(NewUserDTO newUserData);
```
Comment (zmbq) [inline sqe-http-api/Services/UserService.cs]: Ooooh. It's the `NewUserRequestDTO` ! You should rename the class, then.

File: sqe-http-api/Services/UserService.cs [hunk 19]
```diff
@@ -104,16 +122,110 @@ public UserDTO GetCurrentUser()
...
+        /// <returns>Returns a UserDTO for the newly created user account.</returns>
+        public async Task<UserDTO> CreateNewUserAsync(NewUserDTO newUserData)
+        {
+            // Ask the repo to create the new user
+            var createdUser = await _userRepository.CreateNewUserAsync(newUserData.userName, newUserData.email, newUserData.password,
+                newUserData.forename, newUserData.surname, newUserData.organization);
+            
+            // Email the user
+            // TODO: Add link to web endpoint when we know what that is. Can token be in URL query?
+            const string emailBody = @"
```
Comment (zmbq) [inline sqe-http-api/Services/UserService.cs:216]: This should probably go into another file, a resource maybe.

File: sqe-http-api/Services/UserService.cs [hunk 20]
```diff
@@ -104,16 +122,110 @@ public UserDTO GetCurrentUser()
...
+<html><body>Dear $User,<br>
+<br>
+Thank you for registering with the Scripta Qumranica Electronica research platform.  The token
+to activate your new account is: $Token.<br>
```
Comment (zmbq) [inline sqe-http-api/Services/UserService.cs:220]: Also, don't place tokens in the email, you should put a link that user clicks (I understand you used this for debugging purposes)

File: sqe-http-api/Services/UserService.cs [hunk 21]
```diff
@@ -104,16 +122,110 @@ public UserDTO GetCurrentUser()
...
+            var userInfo = await _userRepository.RequestResetForgottenPasswordAsync(email);
+            if (userInfo == null) // Silently return on error
+                return;
+            
+            // Email the user
+            // TODO: Add link to web endpoint when we know what that is. Can token be in URL query?
+            const string emailBody = @"
+<html><body>Dear $User,<br>
+<br>
+Sorry to hear that you have lost your password for Scripta Qumranica Electronica.  You may reset your password with the token: $Token.<br>
```
Comment (zmbq) [inline sqe-http-api/Services/UserService.cs:280]: Same here. A link instead of a token.

File: sqe-http-api/Services/UserService.cs [hunk 22]
```diff
@@ -104,16 +122,110 @@ public UserDTO GetCurrentUser()
...
+        /// <summary>
+        /// Change password from old password to new password for the user's account.
+        /// </summary>
+        /// <param name="user">User object with all information for the user requesting a password change</param>
+        /// <param name="oldPassword">The old password for the user's account</param>
+        /// <param name="newPassword">The new password for the user's account</param>
+        /// <returns></returns>
+        public async Task ChangePasswordAsync(UserInfo user, string oldPassword, string newPassword)
+        {
+            await _userRepository.ChangePasswordAsync(user, oldPassword, newPassword);
```
Comment (zmbq) [inline sqe-http-api/Services/UserService.cs:302]: Do we want to email someone when their password has been changed? A lot of services do that.

**Review by zmbq — CHANGES_REQUESTED** [review]

File: data-access/UserRepository.cs (hunk 9 above)
Comment (Bronson-Brown-deVost) [reply data-access/UserRepository.cs]: Good point, though I disagree about being able to login.  The API should not log in valid but unactivated users (i.e., send them a JWT), it should instead send a response letting them know they are authenticated, but not authorized, and need still need to activate (I guess HTTP error 403 applies here).  We can add an endpoint for a request to resend the activation email (user provides the email address). Let me know what you think, and I can add this.

File: data-access/Queries/User.cs (hunk 6 above)
Comment (Bronson-Brown-deVost) [reply data-access/Queries/User.cs]: Yep, thanks.

File: data-access/Queries/User.cs (hunk 7 above)
Comment (Bronson-Brown-deVost) [reply data-access/Queries/User.cs]: You're right.  I was think authenticated in terms of the Email address being authenticated.  I have changed this to activated in the database and in the API (along with renaming of classes/methods).

Comment (Bronson-Brown-deVost) [reply data-access/Queries/User.cs]: Done.

File: data-access/Queries/User.cs (hunk 6 above)
Comment (Bronson-Brown-deVost) [reply data-access/Queries/User.cs]: Done.

File: data-access/Queries/User.cs (hunk 8 above)
Comment (Bronson-Brown-deVost) [reply data-access/Queries/User.cs:119]: See below. 
Done.

File: data-access/UserRepository.cs (hunk 10 above)
Comment (Bronson-Brown-deVost) [reply data-access/UserRepository.cs]: I've removed all the "clean old token" code, it does not belong in the API.  The database itself deletes tokens once they have expired (I have set the expiration time to two days).

File: sqe-http-api/Controllers/UserController.cs (hunk 12 above)
Comment (Bronson-Brown-deVost) [reply sqe-http-api/Controllers/UserController.cs]: Thanks, I have updated the code and the description in our Website API Google Doc.

File: sqe-http-api/Controllers/UserController.cs (hunk 13 above)
Comment (Bronson-Brown-deVost) [reply sqe-http-api/Controllers/UserController.cs]: Thanks.

Done.

File: sqe-http-api/Controllers/UserController.cs (hunk 14 above)
Comment (Bronson-Brown-deVost) [reply sqe-http-api/Controllers/UserController.cs]: Thanks, I have updated the code and the description in our Website API Google Doc.

File: sqe-http-api/DTOs/User.cs (hunk 15 above)
Comment (Bronson-Brown-deVost) [reply sqe-http-api/DTOs/User.cs]: Renamed according to comment below.

Done.

File: sqe-http-api/DTOs/User.cs (hunk 16 above)
Comment (Bronson-Brown-deVost) [reply sqe-http-api/DTOs/User.cs]: This was a request DTO, not a response DTO.  The DTO has been renamed according to the comment below.

Done.

File: sqe-http-api/Services/UserService.cs (hunk 18 above)
Comment (Bronson-Brown-deVost) [reply sqe-http-api/Services/UserService.cs]: Thanks.  It is renamed now.

Done.

File: sqe-http-api/Services/UserService.cs (hunk 20 above)
Comment (Bronson-Brown-deVost) [reply sqe-http-api/Services/UserService.cs:220]: Agreed (see the TODO in line 137).

File: sqe-http-api/Services/UserService.cs (hunk 21 above)
Comment (Bronson-Brown-deVost) [reply sqe-http-api/Services/UserService.cs:280]: Agreed (~line 177).

File: sqe-http-api/Services/UserService.cs (hunk 22 above)
Comment (Bronson-Brown-deVost) [reply sqe-http-api/Services/UserService.cs:302]: That is a nice idea.  I have added that.

Done.

File: sqe-http-api/Helpers/SystemEmailer.cs (hunk 17 above)
Comment (Bronson-Brown-deVost) [reply sqe-http-api/Helpers/SystemEmailer.cs:22]: Its not a bad idea, especially if we will be sending several types of emails, then we could use a standardized template.  I would like to do this, but I will push it off for now.

File: sqe-http-api/Services/UserService.cs (hunk 19 above)
Comment (Bronson-Brown-deVost) [reply sqe-http-api/Services/UserService.cs:216]: Yeah, I like the idea you mentioned above to use a Razor template, so I am going to push this off for now till we have that set up.

File: data-access/UserRepository.cs (hunk 9 above)
Comment (zmbq) [reply data-access/UserRepository.cs]: The solution of returning 200 with no token is good enough.

File: data-access/Models/ScrollModels.cs [hunk 23]
```diff
@@ -12,7 +12,7 @@ public class Edition
         public bool Locked { get; set; }
         public bool IsPublic {get; set;}
         public DateTime? LastEdit { get; set; }
-        public User Owner { get; set; }
+        public UserToken Owner { get; set; }
```
Comment (zmbq) [inline data-access/Models/ScrollModels.cs:15]: Why are we returning the owner's token? I see no good use cases for this, just bad ones.

File: data-access/Models/ScrollModels.cs [hunk 24]
```diff
@@ -32,7 +32,7 @@ public class Permission
 
     public class Share
     {
-        public User User { get; set; }
+        public UserToken UserToken { get; set; }
```
Comment (zmbq) [inline data-access/Models/ScrollModels.cs:35]: Same question here - why return a Token?

File: data-access/Models/UserModels.cs [hunk 25]
```diff
@@ -7,8 +7,33 @@ public class User
...
+        public string Organization { get; set; }
+        public string Email { get; set; }
+    }
+
+    public class UserEmail : User
+    {
+        public string Forename { get; set; }
+        public string Surname { get; set; }
+        public string Email { get; set; }
+    }
```
Comment (zmbq) [inline data-access/Models/UserModels.cs:36]: I'm beginning to think there are too many user DTOs here, making it cumbersome. Let's not change it for now, but let's not add any new DTOs but rather organize these (or simplify them).

File: data-access/Queries/User.cs [hunk 26]
```diff
@@ -1,31 +1,64 @@
...
+WHERE $Where";
 
-        public User CreateModel()
+        /// <summary>
+        /// Returns a formatted query on the user table
+        /// </summary>
+        /// <param name="columns">Names of the columns to be retrieved (in snake_case)</param>
+        /// <param name="where">Names of the where parameters (in snake_case)</param>
+        /// <returns></returns>
+        public static string GetQuery(List<string> columns, List<string> where)
```
Comment (zmbq) [inline data-access/Queries/User.cs]: Now I understand the ToPascalCase. I think this is really overkill. Simply return all the columns and pick the ones you decide to copy to the DTO. We will never face performance issues due to queries of the user table.

File: data-access/UserRepository.cs [hunk 27]
```diff
@@ -39,19 +47,45 @@ public UserRepository(IConfiguration config) : base(config)
...
-                var results = await connection.QueryAsync<UserQueryResponse>(sql, new
+                return await connection.QuerySingleAsync<UserLoginResponse>(LoginQuery.GetQuery, new
                 {
                     UserName = userName,
                     Password = password
                 });
+            }
+        }
+
+        public async Task<DetailedUserToken> GetDetailedUserById(UserInfo userInfo)
```
Comment (zmbq) [inline data-access/UserRepository.cs]: OK, now there are too many repository methods, too, and probably controller endpoints. Why don't we have just two user DTOs - one with all the fields and the other with all the fields *and* the token, and we always return all of them? Everything will be simpler - the API documentation, the queries, the repository, the service, the controller.

File: sqe-http-api/DTOs/User.cs [hunk 28]
```diff
@@ -20,10 +20,11 @@ public class UserDTO
 
     public class LoginResponseDTO : UserDTO
     {
-        public string token { get; set; }
+        public string token { get; set; } // Used with successful login to activated account
```
Comment (zmbq) [inline sqe-http-api/DTOs/User.cs:23]: Isn't this really the *only place* where we return a token? And nowhere else? So we don't really need a token in the UserDTO - just all the fields?

**Review by zmbq — CHANGES_REQUESTED** [review]

**Comment by Bronson-Brown-deVost** [comment]
Ok, I think this PR has been worked enough for now.  So let's pull it in.
I have tested quite a few user scenarios, and all seems to be working as expected.  I would like to write tests for these possibilities to make sure (but we will need to review and pull the integration-tests branch).

**Review by zmbq — APPROVED** [review]

**Pull request merged**